src/Engine/ProtoCore/FFI/CLRObjectMarshler.cs

```diff
@@ -304,6 +316,38 @@ private static Type GetElementType(Type expectedCLRType)
             return elementType;
         }
 
+        #region DS_ARRAY_TO_CS_DICTIONARY
+
+        protected object ToDictionary(StackValue dsObject, ProtoCore.Runtime.Context context, Interpreter dsi, System.Type keyType, System.Type valueType)
+        {
+            if (!dsObject.IsArray)
+            {
+                return null;
+            }
+
+            var dict = (IDictionary)Activator.CreateInstance(typeof(Dictionary<,>).MakeGenericType(keyType, valueType)); 
+            var dsKeys = ArrayUtils.GetKeys(dsObject, dsi.runtime.RuntimeCore);
+
+            foreach (var dsKey in dsKeys)
+            {
+                var key = primitiveMarshaler.UnMarshal(dsKey, context, dsi, keyType);
+                if (key == null || !keyType.IsAssignableFrom(key.GetType()))
```

Comment:
Do we need to check IsAssignableFrom becuase of string failure or there are some other scenarios where this may not work? I assume that if the type of key is passed for marshaling, then marshaler should have handled it correctly.